src/shogun/lib/type_case.h

```diff
@@ -0,0 +1,152 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Gil Hoben
+ */
+
+#ifndef SHOGUN_TYPE_CASE_H
+#define SHOGUN_TYPE_CASE_H
+
+#include <typeindex>
+#include <unordered_map>
+
+#include <shogun/lib/any.h>
+#include <shogun/lib/type_list.h>
+
+using namespace shogun;
+
+// here we define a list of all types that are implemented in shogun
+// the head of the list returns the type
+typedef Types<
+    bool, char, int8_t, uint8_t, int16_t, uint16_t, int32_t, uint32_t, int64_t,
+    uint64_t, float32_t, float64_t, floatmax_t>::type AllTypes;
+
+namespace shogun
+{
+	// here we make types switchable
+	enum class TYPE
+	{
+		PT_BOOL = 1,
+		PT_CHAR = 2,
+		PT_INT8 = 3,
+		PT_UINT8 = 4,
+		PT_INT16 = 5,
+		PT_UINT16 = 6,
+		PT_INT32 = 7,
+		PT_UINT32 = 8,
+		PT_INT64 = 9,
+		PT_UINT64 = 10,
+		PT_FLOAT32 = 11,
+		PT_FLOAT64 = 12,
+		PT_FLOATMAX = 13,
+		PT_SGOBJECT = 14,
+		PT_COMPLEX128 = 15,
+		PT_UNDEFINED = 16
+	};
+} // namespace shogun
+
+template <typename T>
+struct sg_primitive_type
+{
+};
+
+#define SG_PRIMITIVE_TYPE(T, ptype)                                            \
```

Comment:
make sure to udef all the macros after using them. We don't want to pollute the scope too much